File: backend/app/utils/family_id.py
import re
import secrets
from typing import Set

_d = [
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    [1, 2, 3, 4, 0, 6, 7, 8, 9, 5],
    [2, 3, 4, 0, 1, 7, 8, 9, 5, 6],
    [3, 4, 0, 1, 2, 8, 9, 5, 6, 7],
    [4, 0, 1, 2, 3, 9, 5, 6, 7, 8],
    [5, 9, 8, 7, 6, 0, 4, 3, 2, 1],
    [6, 5, 9, 8, 7, 1, 0, 4, 3, 2],
    [7, 6, 5, 9, 8, 2, 1, 0, 4, 3],
    [8, 7, 6, 5, 9, 3, 2, 1, 0, 4],
    [9, 8, 7, 6, 5, 4, 3, 2, 1, 0],
]

_p = [
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    [1, 5, 7, 6, 2, 8, 3, 0, 9, 4],
    [5, 8, 0, 3, 7, 9, 6, 1, 4, 2],
    [8, 9, 1, 6, 0, 4, 3, 5, 2, 7],
    [9, 4, 5, 3, 1, 2, 6, 8, 7, 0],
    [4, 2, 8, 6, 5, 7, 3, 9, 0, 1],
    [2, 7, 9, 3, 8, 0, 6, 4, 1, 5],
    [7, 0, 4, 6, 9, 1, 3, 2, 5, 8],
]

_inv = [0, 4, 3, 2, 1, 5, 6, 7, 8, 9]


def verhoeff_check_digit(num: str) -> str:
    """Calculate the Verhoeff check digit for a string of numeric digits."""
    c = 0
    for i, ch in enumerate(reversed(num)):
        c = _d[c][_p[(i + 1) % 8][int(ch)]]
    return str(_inv[c])


def verhoeff_valid(num: str) -> bool:
    """Validate a numeric string containing its check digit using Verhoeff algorithm."""
    if not num or not num.isdigit():
        return False
    c = 0
    for i, ch in enumerate(reversed(num)):
        c = _d[c][_p[i % 8][int(ch)]]
    return c == 0


def new_family_id(district_code: int, year: int, taken: Set[str] = None) -> str:
    """
    Generate a 12-digit formatted Family ID: GJ-DD-YY-SSSSSSS-C
    Example: GJ-07-26-4831927-1
    """
    if taken is None:
        taken = set()

    while True:
        serial = f"{secrets.randbelow(10**7):07d}"
        body = f"{district_code:02d}{year % 100:02d}{serial}"
        check = verhoeff_check_digit(body)
        fid = f"GJ-{body[:2]}-{body[2:4]}-{body[4:]}-{check}"
        if fid not in taken:
            return fid


def is_valid_family_id(fid: str) -> bool:
    """
    Validate a Family ID string format and Verhoeff check digit.
    Accepts formats with or without hyphens or spaces.
    """
    if not fid:
        return False

    cleaned = fid.upper().replace(" ", "").replace("-", "")
    if not cleaned.startswith("GJ"):
        return False

    digits = cleaned[2:]
    if len(digits) != 12:
        return False

    return verhoeff_valid(digits)

File: frontend/src/utils/familyId.ts
const d: number[][] = [
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
  [1, 2, 3, 4, 0, 6, 7, 8, 9, 5],
  [2, 3, 4, 0, 1, 7, 8, 9, 5, 6],
  [3, 4, 0, 1, 2, 8, 9, 5, 6, 7],
  [4, 0, 1, 2, 3, 9, 5, 6, 7, 8],
  [5, 9, 8, 7, 6, 0, 4, 3, 2, 1],
  [6, 5, 9, 8, 7, 1, 0, 4, 3, 2],
  [7, 6, 5, 9, 8, 2, 1, 0, 4, 3],
  [8, 7, 6, 5, 9, 3, 2, 1, 0, 4],
  [9, 8, 7, 6, 5, 4, 3, 2, 1, 0],
];

const p: number[][] = [
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
  [1, 5, 7, 6, 2, 8, 3, 0, 9, 4],
  [5, 8, 0, 3, 7, 9, 6, 1, 4, 2],
  [8, 9, 1, 6, 0, 4, 3, 5, 2, 7],
  [9, 4, 5, 3, 1, 2, 6, 8, 7, 0],
  [4, 2, 8, 6, 5, 7, 3, 9, 0, 1],
  [2, 7, 9, 3, 8, 0, 6, 4, 1, 5],
  [7, 0, 4, 6, 9, 1, 3, 2, 5, 8],
];

const inv: number[] = [0, 4, 3, 2, 1, 5, 6, 7, 8, 9];

/** Calculate Verhoeff check digit for numeric string */
export function verhoeffCheckDigit(num: string): string {
  let c = 0;
  const reversed = num.split('').reverse();
  for (let i = 0; i < reversed.length; i++) {
    c = d[c][p[(i + 1) % 8][parseInt(reversed[i], 10)]];
  }
  return String(inv[c]);
}

/** Validate numeric string with its check digit */
export function verhoeffValid(num: string): boolean {
  if (!num || !/^\d+$/.test(num)) return false;
  let c = 0;
  const reversed = num.split('').reverse();
  for (let i = 0; i < reversed.length; i++) {
    c = d[c][p[i % 8][parseInt(reversed[i], 10)]];
  }
  return c === 0;
}

/** Format raw input to standard GJ-DD-YY-SSSSSSS-C */
export function formatFamilyId(input: string): string {
  const cleaned = input.toUpperCase().replace(/[^A-Z0-9]/g, '');
  
  if (!cleaned.startsWith('GJ')) {
    return cleaned;
  }
  
  const digits = cleaned.slice(2);
  let result = 'GJ';

  if (digits.length > 0) result += '-' + digits.slice(0, 2);
  if (digits.length > 2) result += '-' + digits.slice(2, 4);
  if (digits.length > 4) result += '-' + digits.slice(4, 11);
  if (digits.length > 11) result += '-' + digits.slice(11, 12);

  return result;
}

/** Validate complete Family ID string */
export function validateFamilyId(fid: string): { isValid: boolean; error?: string } {
  if (!fid) return { isValid: false, error: 'Family ID is required' };
  
  const cleaned = fid.toUpperCase().replace(/[^A-Z0-9]/g, '');
  if (!cleaned.startsWith('GJ')) {
    return { isValid: false, error: 'Family ID must start with GJ' };
  }

  const digits = cleaned.slice(2);
  if (digits.length !== 12) {
    return { isValid: false, error: 'Family ID must have 12 numeric digits' };
  }

  if (!verhoeffValid(digits)) {
    return { isValid: false, error: 'Invalid check digit (Verhoeff verification failed)' };
  }

  return { isValid: true };
}
